Foundation/NSOperation.swift

```diff
@@ -71,82 +140,392 @@ public class NSOperation : NSObject {
-    case VeryLow
-    case Low
-    case Normal
-    case High
-    case VeryHigh
-}
-
+/*  The NSBlockOperation class is a concrete subclass of NSOperation that manages the concurrent execution
+    of one or more blocks. You can use this object to execute several blocks at once without having to 
+    create separate operation objects for each. When executing more than one block, the operation itself 
+    is considered finished only when all blocks have finished executing. */
 public class NSBlockOperation : NSOperation {
     
-    public convenience init(block: () -> Void) {
-        NSUnimplemented()
+    private typealias ExecutionBlock = () -> ()
+    private var _executionBlocks = [ExecutionBlock]()
+    private let _executionBlocksLock = NSLock()
+
+    private let _concurrencyLimitingSemaphore = DispatchSemaphoreBridge(count: NSOperationQueueDefaultMaxConcurrentOperationCount)
+    private let _queue = DispatchQueueBridge(name: "com.apple.queue.NSBlockOperation.\(NSUUID().UUIDString)", type: .Concurrent)
+
+    private let _dispatchGroup = DispatchGroupBridge()
+
+    public override init() {
+        super.init()
+        _dispatchGroup.enter()
+    }
+
+    public convenience init(block: () -> ()) {
+        self.init()
+        addExecutionBlock(block)
+    }
+
+    public override var cancelled: Bool {
+        _stateLock.lock()
+        defer { _stateLock.unlock() }
+
+        return _state == .Cancelled
+    }
+
+    public override func cancel() {
+        _stateLock.lock()
+        defer { _stateLock.unlock() }
+
+        // TODO: Stop the queue.
+
+        _state = .Cancelled
+    }
+
+    public override var executing: Bool {
+        _stateLock.lock()
+        defer { _stateLock.unlock() }
+
+        return _state == .Executing
+    }
+
+    public override var finished: Bool {
+        _stateLock.lock()
+        defer { _stateLock.unlock() }
+
+        return _state == .Finished || _state == .Cancelled
+    }
+
+    public override var asynchronous: Bool {
+        return true
     }
     
-    public func addExecutionBlock(block: () -> Void) {
-        NSUnimplemented()
+    public func addExecutionBlock(block: () -> ()) {
+
+        _stateLock.lock()
+        defer { _stateLock.unlock() }
+
+        guard _state != .Executing else {
+            fatalError("Cannot add a block if the operation is currently executing.")
+        }
+        
+        guard _state != .Finished && _state != .Cancelled else {
+            fatalError("Cannot add a block if the operation has already finished.")
+        }
+        
+        _executionBlocksLock.lock()
+        defer { _executionBlocksLock.unlock() }
+        _executionBlocks.append(block)
     }
 
-    public var executionBlocks: [() -> Void] {
-        NSUnimplemented()
+    public var executionBlocks: [() -> ()] {
+        _executionBlocksLock.lock()
+        defer { _executionBlocksLock.unlock() }
+        return _executionBlocks
+    }
+
+    public override func start() {
```

Comment:
start should actually be implemented in the abstract base class; since most operations will be subclasses of NSOperation